integration-tests/src/test/java/com/datastax/oss/driver/core/session/ListenersIT.java

```diff
@@ -90,71 +94,102 @@ public void should_inject_session_in_listeners() throws Exception {
 
       assertThat(nodeCaptor2.getValue().getEndPoint())
           .isEqualTo(SIMULACRON_RULE.getContactPoints().iterator().next());
 
-      verify(schemaListener1).onSessionReady(session);
-      verify(schemaListener2).onSessionReady(session);
-
-      verify(requestTracker1).onSessionReady(session);
-      verify(requestTracker2).onSessionReady(session);
-
-      assertThat(MyNodeStateListener.onSessionReadyCalled).isTrue();
-      assertThat(MyNodeStateListener.onUpCalled).isTrue();
-
-      assertThat(MySchemaChangeListener.onSessionReadyCalled).isTrue();
-
-      assertThat(MyRequestTracker.onSessionReadyCalled).isTrue();
+      verify(schemaListener1, timeout(500).times(1)).onSessionReady(session);
+      verify(schemaListener2, timeout(500).times(1)).onSessionReady(session);
+
+      verify(requestTracker1, timeout(500).times(1)).onSessionReady(session);
+      verify(requestTracker2, timeout(500).times(1)).onSessionReady(session);
+
+      assertThat(
+              Uninterruptibles.awaitUninterruptibly(
+                  MyNodeStateListener.onSessionReadyLatch, 500, TimeUnit.MILLISECONDS))
+          .isTrue();
+      assertThat(
+              Uninterruptibles.awaitUninterruptibly(
+                  MyNodeStateListener.onUpLatch, 500, TimeUnit.MILLISECONDS))
+          .isTrue();
+
+      assertThat(
+              Uninterruptibles.awaitUninterruptibly(
+                  MySchemaChangeListener.onSessionReadyLatch, 500, TimeUnit.MILLISECONDS))
+          .isTrue();
+
+      assertThat(
+              Uninterruptibles.awaitUninterruptibly(
+                  MyRequestTracker.onSessionReadyLatch, 500, TimeUnit.MILLISECONDS))
+          .isTrue();
     }
 
-    verify(nodeListener1).close();
-    verify(nodeListener2).close();
-
-    verify(schemaListener1).close();
-    verify(schemaListener2).close();
-
-    verify(requestTracker1).close();
-    verify(requestTracker2).close();
-
-    assertThat(MyNodeStateListener.closeCalled).isTrue();
-    assertThat(MySchemaChangeListener.closeCalled).isTrue();
-    assertThat(MyRequestTracker.closeCalled).isTrue();
+    verify(nodeListener1, timeout(500).times(1)).close();
+    verify(nodeListener2, timeout(500).times(1)).close();
+
+    verify(schemaListener1, timeout(500).times(1)).close();
+    verify(schemaListener2, timeout(500).times(1)).close();
+
+    verify(requestTracker1, timeout(500).times(1)).close();
+    verify(requestTracker2, timeout(500).times(1)).close();
+
+    assertThat(
+            Uninterruptibles.awaitUninterruptibly(
+                MyNodeStateListener.closeLatch, 500, TimeUnit.MILLISECONDS))
+        .isTrue();
+    assertThat(
+            Uninterruptibles.awaitUninterruptibly(
+                MySchemaChangeListener.closeLatch, 500, TimeUnit.MILLISECONDS))
+        .isTrue();
+    assertThat(
+            Uninterruptibles.awaitUninterruptibly(
+                MyRequestTracker.closeLatch, 500, TimeUnit.MILLISECONDS))
+        .isTrue();
```

Comment:
Do we think the close() methods are a problem yet?  This seems like a lot of code to address that case.  I guess I'm wondering if a single timeout check (or even a CountDownLatch perhaps) would do the job here.

I say all that having not actually looked at the relevant code yet.

Reply:
close for all listeners is called before the session's closeFuture is completed so you are right and there were no issues with how close was being checked previously in the test, I'll revert those changes